### ISSUE TITLE
Write query performance tests

### ISSUE THREAD

**Issue by jamesaoverton** [ontodev/valve.rs#26]
The primary goal of the test data set in #16 is to test VALVE load performance. I'd also like to use it to test query performance. These tests will be similar to what we're doing in [sprocket](https://github.com/ontodev/sprocket), and can be written in Python. I'm mostly interested in JSON output. Given a table T, all these tests should account for rows in both T and T_conflict. I want to measure how long it takes to:

1. count all rows
2. count all rows matching a WHERE LIKE condition on a column that is not indexed
3. return the last 100 rows as JSON with simple values: `{"row_number": 99900, "col1": "val1", ...}`
4. return the last 100 rows as JSON with "cell" objects: `{"row_number": 99900, "col1": {"value": "val1", "messages": []}, ...}`
5. return the first 100 rows that have errors as JSON with cells
6. export the message table for T

**Comment by lmcmicu** [comment]
Checked with @jamesaoverton to clarify Question 1, and the idea here is to check whether querying from the beginning vs. the end of the table makes a difference (e.g. `select ... limit 100 offset 10000`)

**Comment by jamesaoverton** [comment]
This is great!

My feeling is that something like 10 runs is enough, but 10,000 rows is not enough. I have tables with several million rows that I want to work with quickly. The tests should run over 100,000 or maybe a million rows. I would like to see a summary without all the hundreds of result rows.

for beautifying JSON I prefer `jq`.

**Comment by lmcmicu** [comment]
Unfortunately the data generation script is not the most efficient. I'll kick off generation of 100,000 rows and see how long it takes. 1,000,000 rows will not be possible with the current script, but it might be possible to optimize it.

**Comment by lmcmicu** [comment]
(BTW, thanks for suggesting `dq`. Although I used duckduckgo above, I actually normally like to use pyton's **json** module for this, in the following way: `COMMAND | python3 -m json.tool`. The output is usually pretty nice.)

Anyway to help with the performance issue, I've made a small modification to the data generation script. It still isn't the most efficient, but the modification is enough to allow us to generate 100,000 rows in a not-completely-unreasonable (given that we only have to do it once) amount of time: ~1hr.

I also changed the query_tests.py script so that the actual rows are not included in the results, as @jamesaoverton requested.

Unfortunately the validation is taking a loooong time to run. It seems to me that the main culprit is the foreign key check which needs to be done while loading some of the tables. The tables without foreign keys load pretty quickly even at 100,000 rows. The tables with foreign keys take forever. One table took almost three hours. There are about three tables like that one! The load has been running since this morning and it is now 3:30PM Munich time.

Anyway I am just going to wait for it to finish and then I will make sure not to delete that database. After it is done I will then run:
```
test/perf_test_data/query_tests/query_tests.py --vacuum 10 table2 bar a 100 99900 build/valve_perf.db | python3 -m json.tool
```
and post the results.

**Comment by jamesaoverton** [comment]
I'm thinking of these query performance checks as distinct from the validation tests in #24. To test validation and loading, 10,000 rows across several tables is good. To test queries (this issue) we only need one table, but it has to be bigger. That one table could be from #24 but it doesn't have to be.

**Comment by lmcmicu** [comment]
Agreed that these tests are separate and do not need to be linked. The problem is just that I don't have a table that big to test on and don't have a way to generate one other than to use valve. I didn't realise it would take quite this long to load, but now that it's started I'd rather just let it run to completion. It's just running in the background now (I'm not actively monitoring it, just occasionally). Hopefully it will be only another few hours. Otherwise I'll look into another way to do it tomorrow.

**Comment by lmcmicu** [comment]
Your last comment made me realise, though, that I could modify `table.tsv` and `column.tsv` to eliminate all but the one table I want to load for the query test, so no foreign key validation should need to be performed. It might still take a little while to load that one table but hopefully not anywhere near the time it was taking to load all six tables

**Comment by lmcmicu** [comment]
The only potential downside is that without the other tables there will be fewer errors, which will potentially make the "first 100 rows that have errors as JSON with cells" test less meaningful. Anyway we'll see.

**Comment by lmcmicu** [comment]
Ok, I used a table without a lot of foreign, tree, or under keys defined on it, and since there weren't enough datatype errors in the table I just manually added about 120 or so. Attached is a similar JSON to what I uploaded yesterday, but this table has 100,000 rows in it.
[output_pretty.txt](https://github.com/ontodev/valve.rs/files/10149211/output_pretty.txt)

**Comment by lmcmicu** [comment]
Ok here is some better data. The table, `table1`, has 100,000 rows. 2,476 of these rows have metadata. In addition there are 2,541 rows in `table1_conflict`.

Note that specifying an offset of N does _not_ mean (as I originally thought) to query the table beginning from row N. Rather, it means, run the query on the whole table, and then skip the first N rows from the results that are returned. With that in mind I specified an offset of 2000 by running the script with the following command line options:
```
test/perf_test_data/query_tests/query_tests.py --vacuum 10 table1 base a 100 2000 build/valve_perf.db | python3 -m json.tool | tee output_pretty.txt
```
Here are the results: 
[output_pretty.txt](https://github.com/ontodev/valve.rs/files/10153223/output_pretty.txt)

**Comment by lmcmicu** [comment]
After getting some feedback from @jamesaoverton about precisely what the contents of these tests should be, I produced the following Excel spreadsheet with the appropriate statistics. Note that some of the tests included here are related to #27
[Schema comparison.xlsx](https://github.com/ontodev/valve.rs/files/10201641/Schema.comparison.xlsx)
